### PR TITLE
Support multiple binds and consistency updates

### PR DESCRIPTION
**Pull request by sethvargo**
This PR does a few things, primarily because I had to re-arrange some things to fix GH-27.

1. **Parallelize tests** - the tests now run in parallel. On the plane last night, they ran in ~8s. Today they are running at ~2s on my home Internet. That's a 5000% decrease in test time 😄! I had to add a few helpers to make sure tests wouldn't step on each other.

1. **Support multiple bindings for region, zone, and instance group** - this fixes GH-27 and provides users the ability to bind to multiple regions, zones, and/or instance groups instead of a single one. If a GCE instance exists in any of the given values, it will authenticate successfully. 

    Given the existing structure of the backend, it was very difficult to test this new logic in an independent way. I built an "authenticator" interface that makes the GCP API calls. For testing, those API calls can be stubbed with arbitrary values to test exclusively the logic. There's now a test for every edge case using these stubbed values that run very quickly to test our logic.

    The old schema is preserved and automatically upgraded the first time a role is accessed. This is inline with the recommendation from the Vault team on how to handle such upgrades. 

1. **Warning passthrough** - theres now a way to pass warnings up from the helper methods, which is helpful during decoding, etc.

1. **Validate input fields** - this is something Vault is doing sporadically, but I added it here everywhere. Nothing is more frustrating than when you accidentally mistype a field (like `vault write auth/gcp/config cradentials=...`) and Vault happily accepts that. Fields that don't match the schema will now inform the user of their error, hopefully saving valuable debugging time.

1. **Reduce the number of API calls** - there are some optimizations to reduce the number of API calls, including server-side filter and client-side string parsing. For example, given a zone, we can **always** calculate the region from the zone using string parsing; we don't have to incur an API. The zone/regions spec specifically documents this format, so we are okay to use it. On average, this saves two API calls per (GCE) authentication.

🚧 This will require a docs update in the main Vault repo. I can take on that work once we are close to merging this.

### CODE REVIEW

File: plugin/path_role.go [hunk 1]
```diff
@@ -7,7 +7,9 @@ import (
 	"strings"
 	"time"
 
+	"github.com/fatih/structs"
```
Comment (emilymye) [inline plugin/path_role.go]: We're trying to avoid using third-party libraries that belong to one person like this as much as possible. Also, I think @jefferai mentioned Vault was moving specifically away from using structs for converting the response and just writing it out instead. 

Otherwise, this requires updates to deps

File: plugin/path_role.go [hunk 2]
```diff
@@ -261,38 +282,22 @@ func (b *GcpAuthBackend) pathRoleRead(ctx context.Context, req *logical.Request,
...
-		}
-	}
+	role.Period /= time.Second
+	role.TTL /= time.Second
+	role.MaxTTL /= time.Second
+	role.MaxJwtExp /= time.Second
 
 	return &logical.Response{
-		Data: roleMap,
+		Data: structs.New(role).Map(),
```
Comment (emilymye) [inline plugin/path_role.go]: per previous comment, don't use structs here

File: plugin/authorizer_client_gcp.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
+package gcpauth
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/vault/helper/strutil"
+	compute "google.golang.org/api/compute/v1"
```
Comment (emilymye) [inline plugin/authorizer_client_gcp.go]: nit/note - I don't think you actually need to alias these imports?

File: plugin/authorizer_client_gcp.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	iamSvc     *iam.Service
+}
+
+func (c *gcpClient) InstanceGroups(ctx context.Context, project string, boundInstanceGroups []string) (map[string][]string, error) {
+	// map of zone names to a slice of instance group names in that zone.
+	igz := make(map[string][]string)
+
+	if err := c.computeSvc.InstanceGroups.
+		AggregatedList(project).
+		Fields("items/*/instanceGroups/name").
```
Comment (emilymye) [inline plugin/authorizer_client_gcp.go]: We should probably use [filters](https://cloud.google.com/compute/docs/reference/rest/v1/instances/aggregatedList#filter) for name and zones  instead of having the pagination function?

**Comment by emilymye** [comment]
>Support pagination for long GCP API responses - could be viewed as a bug, but when the list of instance groups was really long, we weren't paginating. This is fixed.

 A little confused - were we listing instance groups?

File: plugin/authorizer_client_gcp.go (hunk 4 above)
Comment (sethvargo) [reply plugin/authorizer_client_gcp.go]: It's not possible due to a bug in the API: https://stackoverflow.com/questions/50452838/google-cloud-compute-instances-aggregatedlist-with-filter-fails

**Comment by sethvargo** [comment]
> A little confused - were we listing instance groups?

Apparently not. My mistake. I've updated the original issue to remove that.

File: plugin/authorizer_client_gcp.go (hunk 4 above)
Comment (sethvargo) [reply plugin/authorizer_client_gcp.go]: We'd also still need the pagination function, even if we used filters. Filters are applied _after_ the response is generated, so you'd still need to paginate, potentially though a bunch of empty pages. For example, if there are 400 results, and you filter, the first 20 pages might be empty, since they _were_ filled prior to filtering.

But yea - you can test trying to filter in the docs online. It's broken. There's an internal ticket about it.

**Comment by sethvargo** [comment]
@emilymye updated per your other feedback

**Comment by emilymye** [comment]
@sethvargo thanks for doing this! It mostly LGTM but I'm going to take a little longer to review this since we actually don't have tests for the GCE auth right now (something I'm actually trying to get set up seperately now - turns out the fact that GCE metadata must be accessed via an instance makes testing hard and slow, who knew)

**Comment by sethvargo** [comment]
@emilymye yea, unfortunately I'm aware. If it helps, I've tested it manually on a few different GCE instances, but it would be great to have an automated way to test stuff

File: plugin/authorizer_client_gcp.go (hunk 4 above)
Comment (emilymye) [reply plugin/authorizer_client_gcp.go]: wow what a BIG OLE BUNCH OF FUN :(
lgtm then
